File: reference/test/reference_matchers.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:reference/reference.dart';

// Extends isMethodCall in package:test/test.dart to support matchers in arguments.
Matcher isMethodCallWithMatchers(String method, {Object? arguments}) {
  return _IsMethodCallWithMatchers(method, arguments);
}

class _IsMethodCallWithMatchers extends Matcher with _DeepEquals {
  const _IsMethodCallWithMatchers(this.method, this.arguments);

  final String method;
  final Object? arguments;

  @override
  bool matches(dynamic item, Map<dynamic, dynamic> matchState) {
    if (item is! MethodCall) return false;
    if (item.method != method) return false;
    return deepEquals(item.arguments, arguments, matchState);
  }

  @override
  Description describe(Description description) {
    return description
        .add('Is a $MethodCall with method name: ')
        .addDescriptionOf(method)
        .add(' and arguments: ')
        .addDescriptionOf(arguments);
  }
}

// Extends isMethodCall in packager:test/test.dart to support matchers in arguments.
Matcher isUnpairedReference(
  String channelName,
  List<Object> creationArguments,
) {
  return _IsUnpairedReference(channelName, creationArguments);
}

class _IsUnpairedReference extends Matcher with _DeepEquals {
  const _IsUnpairedReference(this.channelName, this.creationArguments);

  final String channelName;

  final List<Object> creationArguments;

  @override
  Description describe(Description description) {
    return description
        .add(' Is an $UnpairedReference with channel name: ')
        .addDescriptionOf(channelName)
        .add(' and creation arguments: ')
        .addDescriptionOf(creationArguments);
  }

  @override
  bool matches(
    covariant UnpairedReference item,
    Map<dynamic, dynamic> matchState,
  ) {
    if (item is! UnpairedReference) return false;
    if (item.channelName != channelName) return false;
    return deepEquals(creationArguments, item.creationArguments, matchState);
  }
}

mixin _DeepEquals {
  bool deepEquals(Object? a, Object? b, Map<Object?, Object?> matchState) {
    if (a == b) return true;
    if (b is Matcher) return b.matches(a, matchState);
    if (a is Matcher) return a.matches(b, matchState);
    if (a is List) return b is List && _deepEqualsList(a, b, matchState);
    if (a is Map) return b is Map && _deepEqualsMap(a, b, matchState);
    return false;
  }

  bool _deepEqualsList(
    List<Object?> a,
    List<Object?> b,
    Map<Object?, Object?> matchState,
  ) {
    if (a.length != b.length) return false;
    for (int i = 0; i < a.length; i++) {
      if (!deepEquals(a[i], b[i], matchState)) return false;
    }
    return true;
  }

  bool _deepEqualsMap(
    Map<Object?, Object?> a,
    Map<Object?, Object?> b,
    Map<Object?, Object?> matchState,
  ) {
    if (a.length != b.length) return false;
    for (final Object? key in a.keys) {
      if (!b.containsKey(key) || !deepEquals(a[key], b[key], matchState)) {
        return false;
      }
    }
    return true;
  }
}

File: reference/android/src/test/java/github/penguin/reference/ReferenceMatchers.java
package github.penguin.reference;

import github.penguin.reference.reference.NewUnpairedInstance;
import io.flutter.plugin.common.MethodCall;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

import java.util.List;

@SuppressWarnings("rawtypes")
public class ReferenceMatchers {
  public static Matcher isMethodCall(String method, Object arguments) {
    return new IsMethodCall(method, arguments);
  }

  public static Matcher isUnpairedInstance(String channelName, final Object creationArguments) {
    return new IsUnpairedInstance(channelName, creationArguments);
  }

  private static class IsMethodCall extends TypeSafeMatcher<MethodCall> {
    private final String method;
    private final Object arguments;

    private IsMethodCall(String method, Object arguments) {
      this.method = method;
      this.arguments = arguments;
    }

    private void describe(final String method, final Object arguments, Description description) {
      description
          .appendText(String.format("A %s with method name: ", MethodCall.class.getSimpleName()))
          .appendText(method)
          .appendText(" and arguments: ")
          .appendText(arguments.toString());
    }

    @Override
    public void describeTo(Description description) {
      describe(method, arguments, description);
    }

    @Override
    protected void describeMismatchSafely(MethodCall call, Description mismatchDescription) {
      describe(call.method, call.arguments, mismatchDescription);
    }

    @Override
    protected boolean matchesSafely(final MethodCall call) {
      if (!call.method.equals(method)) return false;
      if (arguments instanceof Matcher) return ((Matcher) arguments).matches(call.arguments);
      return arguments.equals(call.arguments);
    }
  }

  private static class IsUnpairedInstance extends TypeSafeMatcher<NewUnpairedInstance> {
    private final String channelName;
    private final Object creationArguments;

    private IsUnpairedInstance(String channelName, Object creationArguments) {
      this.channelName = channelName;
      this.creationArguments = creationArguments;
    }

    private void describe(
        final String channelName,
        final Object creationArguments,
        Description description) {
      description
          .appendText(
              String.format(" An %s with channelName: ", NewUnpairedInstance.class.getSimpleName()))
          .appendText(channelName)
          .appendText(" and creation arguments: ")
          .appendText(creationArguments != null ? creationArguments.toString() : null);
    }

    @Override
    public void describeTo(Description description) {
      describe(channelName, creationArguments, description);
    }

    @Override
    protected void describeMismatchSafely(
        NewUnpairedInstance unpairedInstance, Description mismatchDescription) {
      describe(
          unpairedInstance.channelName,
          unpairedInstance.creationArguments,
          mismatchDescription);
    }

    @Override
    protected boolean matchesSafely(NewUnpairedInstance unpairedInstance) {
      if (!channelName.equals(unpairedInstance.channelName)) return false;
      if (creationArguments instanceof List) {
        return creationArguments.equals(unpairedInstance.creationArguments);
      }
      if (creationArguments instanceof Matcher) {
        return ((Matcher) creationArguments).matches(unpairedInstance.creationArguments);
      }
      return creationArguments == unpairedInstance.creationArguments;
    }
  }
}
